### DISCUSSION TITLE
How to obtain PIL.Image and PIL.ImageOps? (Pillow)

### DISCUSSION THREAD

**Discussion by JohnScience** [PyO3/pyo3#3001] [Questions] [answered]
Hi, I'm rewriting a piece of software from Python to Rust.

This piece relies on [`Pillow`](https://pillow.readthedocs.io/en/stable/index.html) library, which is a fork of `PIL`.

For some reason, it does not permit using `PIL.Image` but does permit importing it in Python like this:

```python
from PIL import Image
```

So the following Rust code did not work for me:

```rust
Python::with_gil(|py| {
    // ...
    let (Image, ImageOps) = {
        let PIL = py.import("PIL")?;
        (
            PIL.getattr("Image")?,
            PIL.getattr("ImageOps")?,
        )
    };
    // ...
}
```

I tried obtaining them via a `.dict()` from `PIL` module...

```rust
Python::with_gil(|py| {
    // ...
    let (Image, ImageOps) = {
        let PIL = py.import("PIL")?;
        let dict = PIL.dict();
        (
            dict.get_item("Image").ok_or(PyImportError::new_err("module 'PIL' has no attribute 'Image'"))?,
            dict.get_item("ImageOps").ok_or(PyImportError::new_err("module 'PIL' has no attribute 'ImageOps'"))?,
        )
    };
    // ...
}
```

but I failed again.

I guess I should somehow use [`importlib`](https://docs.python.org/3/library/importlib.html#), yet I'm still in search of what I should have done.

**Comment by JohnScience** [comment, marked as answer]
I had a chat with Chat GPT and it proposed me this approach:

```python
import importlib

PIL = importlib.import_module('PIL.Image')
```

and it worked. Then I tried doing it like this:

```python
import PIL.Image
```

and it worked too! Hope it will help anyone!

**Comment by birkenfeld** [comment]
For future reference, please include the error(s) you're getting with such reports. It is much easier to see for use what could be going wrong.
